main.cpp

```diff
@@ -124,379 +124,157 @@ struct IntType;
-    {
-        if (value != nullptr) 
-            *value *= operand;
-
-        return *this;
-    }
+    FloatType& add( float operand );
+    FloatType& subtract( float operand );
+    FloatType& multiply( float operand );
+    FloatType& divide( float operand );
 
-    IntType& multiply(const FloatType& ft);
-    IntType& multiply(const DoubleType& ft);
-    IntType& multiply(const IntType& ft);
-    
-    IntType& divide( int operand )
-    {
-        if (operand != 0)
-        {
-            *value /= operand;    
-        }
-        else
-        {
-            std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl ;
-        }
-        return *this;
-    }
+    operator int() { return static_cast<int>(*value); }
 
-    IntType& divide(const FloatType& ft);
-    IntType& divide(const DoubleType& ft);
-    IntType& divide(const IntType& ft);
+private:
+    float* value = nullptr;
 };
 
-FloatType& FloatType::add(const FloatType& ft)
+FloatType& FloatType::add( float operand )
 {
-    return add(*ft.value);
-}
+    if (value != nullptr)    
+        *value += operand;
 
-FloatType& FloatType::add(const DoubleType& ft)
-{
-    return add(*ft.value);
+    return *this;
 }
 
-FloatType& FloatType::add(const IntType& ft)
+FloatType& FloatType::subtract( float operand )
 {
-    return add(*ft.value);
-}
+    if (value != nullptr)   
+        *value -= operand;
 
-FloatType& FloatType::subtract(const FloatType& ft)
-{
-    return subtract(*ft.value);
+    return *this;
 }
 
-FloatType& FloatType::subtract(const DoubleType& ft)
+FloatType& FloatType::multiply( float operand )
 {
-    return subtract(*ft.value);
-}
+    if (value != nullptr) 
+        *value *= operand;
 
-FloatType& FloatType::subtract(const IntType& ft)
-{
-    return subtract(*ft.value);
+    return *this;
 }
 
-FloatType& FloatType::multiply(const FloatType& ft)
+FloatType& FloatType::divide( float operand )
 {
-    return multiply(*ft.value);
-}
+    if (operand == 0.f)
+        std::cout << "warning: floating point division by zero!" << std::endl; 
 
-FloatType& FloatType::multiply(const DoubleType& ft)
-{
-    return multiply(*ft.value);
-}
+    if (value != nullptr)
+        *value /= operand;
 
-FloatType& FloatType::multiply(const IntType& ft)
-{
-    return multiply(*ft.value);
+    return *this;
 }
 
-FloatType& FloatType::divide(const FloatType& ft)
+struct DoubleType
 {
-    return divide(*ft.value);
-}
+    DoubleType(double _value) : value(new double(_value)) { }
+    ~DoubleType() { delete value; }
 
-FloatType& FloatType::divide(const DoubleType& ft)
-{
-    return divide(*ft.value);
-}
+    DoubleType& add( double operand );
+    DoubleType& subtract( double operand );
+    DoubleType& multiply( double operand );
+    DoubleType& divide( double operand );
 
-FloatType& FloatType::divide(const IntType& ft)
-{
-    return divide(*ft.value);
-}
+    operator int() { return static_cast<int>(*value); }
```

Comment:
This class is `DoubleType`
Why would you create an `operator int()` function?
- [ ] revise